### PR TITLE
UAVCAN Smart Battery Improvements

### PR DESCRIPTION
**Pull request by AlexKlimaj**
**Describe problem solved by this pull request**
This PR is a continuation of @dakejahl 's work to integrate a UAVCAN smart battery.

**Describe your solution**
Improved the UAVCAN battery sensor bridge driver. Un-deprecated BAT_SOURCE to use as a parameter to disable battery_status module when a UAVCAN battery is being used. 

**Test data / coverage**
Flight testing has been done with prototype ARK BMS smart batteries with UAVCAN. Flight testing with a single and dual batteries have been done.
[Dual Battery Flight Test](https://review.px4.io/plot_app?log=e981e85c-466d-4841-9eff-c4094cef5e57)
[Single Battery 25 Minute Flight Test](https://review.px4.io/plot_app?log=f0221744-79fc-4473-be30-932dcd7dbc2e)

image
image
image

**Additional context**
[See the BMS Workgroup.](https://docs.google.com/document/d/1qxKdUSQv-poblcyOBltgV3QO7NZ7xupk6KM-3PkeJGs/edit?usp=sharing)

### CODE REVIEW

**Comment by dagar** [comment]
Looks like we need some system wide configuration to properly configure the smart battery scenario.
Things like skipping batery_status start, making the uavcan smart battery an arming requirement, etc

**Comment by AlexKlimaj** [comment]
I just completed a 25 minute hover test. It ended with the battery cutting out which is all on my side configuring the fuel gauge. 

https://review.px4.io/plot_app?log=807e8661-99b8-4253-a840-e627ff2c02b0

We also need to think about how to support multiple batteries in parallel with this driver.

**Comment by AlexKlimaj** [comment]
Here is a 25 minute flight on a Pixhawk 4 with this driver.

https://review.px4.io/plot_app?log=f0221744-79fc-4473-be30-932dcd7dbc2e

**Comment by AlexKlimaj** [comment]
@dagar Just got a short flight with two smart batteries. With commander and mavlink already setup to handle multiple batteries, this driver works with multiple uavcan packs well.

https://review.px4.io/plot_app?log=e981e85c-466d-4841-9eff-c4094cef5e57
image

**Comment by AlexKlimaj** [comment]
@dagar , @dakejahl and I went through the battery parameters last night and found that the new battery parameters are only loaded if the battery_status module is started. For the time being, I un-depricated BAT_SOURCE and I'm using it to not start battery_status. BAT%D_SOURCE for each battery doesn't really make sense either.

**Comment by AlexKlimaj** [comment]
@dagar Any thoughts?

**Comment by hamishwillee** [comment]
@AlexKlimaj What docs do you have planned for this, if any? 

Without thinking too hard about this, it might be time to create a smart battery page below https://docs.px4.io/master/en/peripherals/
We could talk about what PX4 settings and hardware configuration are required, if any?

Could you kick this off?

**Comment by AlexKlimaj** [comment]
Here is a 20 minute flight with 2 smart batteries. This also tests @davids5's PR https://github.com/PX4/Firmware/pull/14367 on a Pixhawk 4 Mini.

https://review.px4.io/plot_app?log=75f2e613-aec3-4fd4-abd3-dcd1ff9027b5

Voltage
image

Current
image

Remaining
image

Temperature
image

**Comment by dagar** [comment]
> @dagar , @dakejahl and I went through the battery parameters last night and found that the new battery parameters are only loaded if the battery_status module is started. For the time being, I un-depricated BAT_SOURCE and I'm using it to not start battery_status. BAT%D_SOURCE for each battery doesn't really make sense either.

I agree, that looks like it was a mistake.

File: src/drivers/uavcan/sensors/battery.cpp [hunk 1]
```diff
@@ -69,36 +72,72 @@ UavcanBatteryBridge::battery_sub_cb(const uavcan::ReceivedDataStructure<uavcan::
...
+		// Because the measurement was invalid we need to stop integration
+		// and re-initialize with the next valid measurement
+		_last_timestamp = 0;
+		return;
+	}
+
+	// Ignore first update because we don't know dt.
+	if (_last_timestamp != 0) {
+		const float dt = (timestamp - _last_timestamp) / 1e6;
+		// mAh since last loop: (current[A] * 1000 = [mA]) * (dt[s] / 3600 = [h])
```
Comment (dagar) [inline src/drivers/uavcan/sensors/battery.cpp]: Did you consider using the battery lib for this (I haven't looked in detail)?

File: src/drivers/uavcan/sensors/battery.cpp [hunk 2]
```diff
@@ -69,36 +72,72 @@ UavcanBatteryBridge::battery_sub_cb(const uavcan::ReceivedDataStructure<uavcan::
...
-		battery.remaining = 0;
-
-	}
+	sumDischarged(battery.timestamp, battery.current_a);
+	battery.discharged_mah = _discharged_mah;
 
+	battery.remaining = msg.state_of_charge_pct / 100.0f; // between 0 and 1
 	// battery.scale = msg.; // Power scaling factor, >= 1, or -1 if unknown
-	battery.temperature = msg.temperature;
+	battery.temperature = msg.temperature - 273.15f; // Kelvin to Celcius
```
Comment (dagar) [inline src/drivers/uavcan/sensors/battery.cpp]:
```suggestion
	battery.temperature = msg.temperature + CONSTANTS_ABSOLUTE_NULL_CELSIUS; // Kelvin to Celcius
```

File: src/drivers/uavcan/sensors/battery.cpp (hunk 1 above)
Comment (AlexKlimaj) [reply src/drivers/uavcan/sensors/battery.cpp]: I did not consider it. But now that I look, this could inherit from Battery and use those functions.

Comment (AlexKlimaj) [reply src/drivers/uavcan/sensors/battery.cpp]: I don't think it is possible with the current design of the battery lib. Sumdischarged and determineWarning are private members. So I would have to use updateBatteryStatus, but that doesn't have all the fields required. So I could modify the battery lib to work in this situation, or leave these two functions duplicated into this driver. 
Also, UavcanCDevSensorBridgeBase already has its structure of publishing to uorb.

File: src/lib/battery/battery_params_common.c [hunk 3]
```diff
@@ -88,3 +88,20 @@ PARAM_DEFINE_FLOAT(BAT_CRIT_THR, 0.07f);
  * @reboot_required true
  */
 PARAM_DEFINE_FLOAT(BAT_EMERGEN_THR, 0.05f);
+
+/**
+ * This parameter is deprecated. Please use BAT1_SOURCE instead.
```
Comment (dakejahl) [inline src/lib/battery/battery_params_common.c]: Updatethis whole block of info text

Comment (AlexKlimaj) [reply src/lib/battery/battery_params_common.c]: Will do.

**Comment by dakejahl** [comment]
I think you didn't update your submodules, you have checked in a bunch of submodule changes

**Comment by AlexKlimaj** [comment]
> 
> 
> I think you didn't update your submodules, you have checked in a bunch of submodule changes

Yeah, just fixed thanks.

File: src/lib/battery/battery_params_common.c (hunk 3 above)
Comment (AlexKlimaj) [reply src/lib/battery/battery_params_common.c]: Done

**Comment by dakejahl** [comment]
Good to merge @dagar

File: src/lib/battery/battery_params_common.c [hunk 4]
```diff
@@ -88,3 +88,18 @@ PARAM_DEFINE_FLOAT(BAT_CRIT_THR, 0.07f);
...
+ * means that measurements are expected to come from a power module. If the value is set to
+ * 'External' then the system expects to receive mavlink or UAVCAN battery status messages.
+ *
+ * @min 0
+ * @max 1
+ * @value 0 Power Module
+ * @value 1 External
+ * @group Battery Calibration
+ */
+PARAM_DEFINE_INT32(BAT_SOURCE, 0);
```
Comment (julianoes) [inline src/lib/battery/battery_params_common.c]: Why did you move this back? This param was meant to be deprecated.

Comment (AlexKlimaj) [reply src/lib/battery/battery_params_common.c]: You need a way to turn off the battery_status module. @dagar and I discussed that this is the correct use of this parameter. The BAT%_SOURCE params don't make much sense.

Comment (julianoes) [reply src/lib/battery/battery_params_common.c]: Haha, in the current state not, I agree. Unfortunately, I have also some uncommited changes which change it in another way and make `BAT%d_SOURCE` params work. I'll push that soon and we'll have to discuss it all.

**Comment by AlexKlimaj** [comment]
@dagar Ready to merge?

**Comment by julianoes** [comment]
@AlexKlimaj @dagar this will heavily conflict with https://github.com/PX4/Firmware/pull/14490. Could you review my suggestion for what to do with the `BAT_SOURCE` param?

File: src/lib/battery/battery_params_common.c (hunk 4 above)
Comment (AlexKlimaj) [reply src/lib/battery/battery_params_common.c]: So you are using the BAT1_SOURCE like I use BAT_SOURCE here.
https://github.com/PX4/Firmware/pull/14490/files#diff-e5c16c5209413f30e964883f74027b51R351
Do you want me to do that? In the UAVCAN case, there could be more than one battery connected and all I need BAT1_SOURCE to do is turn off the battery_status module.

**Comment by AlexKlimaj** [comment]
> 
> 
> @AlexKlimaj @dagar this will heavily conflict with #14490. Could you review my suggestion for what to do with the `BAT_SOURCE` param?

Merged in #14490 and resolved conflicts.

**Comment by AlexKlimaj** [comment]
@julianoes Got out and flew this PR with your branch merged in today. Everything is looking good. 
Here is the longest flight:
https://review.px4.io/plot_app?log=2df1a307-1adf-463a-9d95-97285343d6d2

File: src/lib/battery/battery.cpp [hunk 5]
```diff
@@ -86,14 +86,14 @@ Battery::Battery(int index, ModuleParams *parent) :
...
 	_param_handles.emergen_thr = param_find("BAT_EMERGEN_THR");
-	_param_handles.source_common = param_find("BAT_SOURCE");
 
 	_param_handles.v_empty_old = param_find("BAT_V_EMPTY");
 	_param_handles.v_charged_old = param_find("BAT_V_CHARGED");
 	_param_handles.n_cells_old = param_find("BAT_N_CELLS");
 	_param_handles.capacity_old = param_find("BAT_CAPACITY");
 	_param_handles.v_load_drop_old = param_find("BAT_V_LOAD_DROP");
 	_param_handles.r_internal_old = param_find("BAT_R_INTERNAL");
+	_param_handles.source_old = param_find("BAT_SOURCE");
```
[the remark this answers is not in the document]
Comment (AlexKlimaj) [reply src/lib/battery/battery.cpp]: Yeah, you are right.

File: src/lib/battery/battery.cpp [hunk 6]
```diff
@@ -93,7 +93,6 @@ Battery::Battery(int index, ModuleParams *parent) :
 	_param_handles.capacity_old = param_find("BAT_CAPACITY");
 	_param_handles.v_load_drop_old = param_find("BAT_V_LOAD_DROP");
 	_param_handles.r_internal_old = param_find("BAT_R_INTERNAL");
-	_param_handles.source_old = param_find("BAT_SOURCE");
```
Comment (julianoes) [inline src/lib/battery/battery.cpp:96]: Sorry, I didn't mean to remove it but leave it as is. Before you had moved it in this file.

Comment (AlexKlimaj) [reply src/lib/battery/battery.cpp:96]: Thanks. Re checked out your branch for the battery lib and pushed.

Comment (julianoes) [reply src/lib/battery/battery.cpp:96]: Great.

**Comment by mrpollo** [comment]
@dagar @julianoes what's missing here? can we merge this in?
